airflow/www/security.py

```diff
@@ -346,10 +346,7 @@ def get_current_user_permissions(self):
         return perms
 
     def current_user_has_permissions(self) -> bool:
-        for role in self.get_user_roles():
-            if role.permissions:
-                return True
-        return False
+        return len(self.get_current_user_permissions()) > 0
```

Comment:
@uranusjr thanks for that feedback on blowing up DB queries. Would this be an acceptable workaround?
Look for roles first, then look for permissions only if necessary?
```suggestion
        for role in self.get_user_roles():
            if role.permissions:
                return True
        return len(self.get_current_user_permissions()) > 0
```

Reply:
alternatively (pardon my ignorance) is there a way for us to cache the user permissions for a sessions so we don't have to make that query twice?

Reply:
It seems like an odd fallback to me as it doesn't make sense in the context of `AirflowSecurityManager`.

I'm not sure where you draw that line. E.g. should `get_accessible_dags` also have a fallback to support subclasses trying to remove the need for roles? What if someone else tried to add another layer instead of removing one?

My thoughts are: If you are redefining where permissions come from (not roles) you should be prepared to implement all of the permissions based methods. I'm not saying changes to make it more subclass-friendly shouldn't happen, however if you need a comment explaining why a fallback is needed that could only happen with a subclass and removal of a core assumption, idk.

Reply:
good point, this logic resonates with me. It's smelly to make implementation choices that cause inefficiency for this default `AirflowSecurityManager` based on "what if people want to subclass this". The purpose of this class is to be the default `AirflowSecurityManager` so sub-classers have to assume the risk / responsibility that methods like this might crop up over time and need to be overridden.

If the goal is to improve the security manager subclassing experience it'd probably be better addressed by pulling out a `BaseAirlfowSecurityManager(abc.ABC)` that clearly spells out the contract of which abstract methods should be overridden to make this work correctly with the rest of airflow and making this AirflowSecurityManager a subclass of that.

that's not really the topic of this PR so I will close it and can open a new PR for implementing a `BaseAirlfowSecurityManager` if you think that's worthwhile.

Reply:
Yeah, having an ABC could make this a lot cleaner 👍. Before you do the work though, it'd be worth pinging @jhtimmins directly to coordinate.